### ISSUE TITLE
[FEATURE] Implement Speaker Voice Embedding and Identification

### ISSUE THREAD

**Issue by davidamacey** [davidamacey/OpenTranscribe#18]
# Speaker Identification and Management System

## Problem Statement
Current speaker diarization only distinguishes between speakers within a single audio file without recognizing the same speaker across different files. We need to implement a robust speaker identification system that can recognize speakers across multiple recordings.

#### `speakers` Table
```sql
CREATE TABLE speakers (
    id SERIAL PRIMARY KEY,
    name VARCHAR(255) NOT NULL,
    user_id INTEGER REFERENCES users(id) ON DELETE CASCADE,
    created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
    updated_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
    metadata JSONB DEFAULT '{}'::jsonb,
    is_active BOOLEAN DEFAULT TRUE,
    UNIQUE(name, user_id)
);

-- Index for faster lookups
CREATE INDEX idx_speakers_user ON speakers(user_id);
CREATE INDEX idx_speakers_name ON speakers USING GIN (name gin_trgm_ops);
```

#### `speaker_embeddings` Table
```sql
CREATE TABLE speaker_embeddings (
    id SERIAL PRIMARY KEY,
    speaker_id INTEGER NOT NULL REFERENCES speakers(id) ON DELETE CASCADE,
    embedding VECTOR(192) NOT NULL,  -- Using pgvector with 192-dimensional embeddings
    source_segment_id INTEGER REFERENCES transcript_segments(id) ON DELETE SET NULL,
    created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
    confidence FLOAT,
    metadata JSONB DEFAULT '{}'::jsonb
);

-- Index for vector similarity search
CREATE INDEX idx_speaker_embeddings_embedding ON speaker_embeddings 
USING ivfflat (embedding vector_cosine_ops) WITH (lists = 100);

-- Index for faster lookups
CREATE INDEX idx_speaker_embeddings_speaker ON speaker_embeddings(speaker_id);
```

### 2. Speaker Embedding Service

```python
import torch
import torchaudio
from pyannote.audio import Model
from pyannote.audio.pipelines import SpeakerDiarization
from typing import List, Dict, Optional
import numpy as np

class SpeakerEmbeddingService:
    def __init__(self, model_name: str = "pyannote/embedding"):
        self.device = "cuda" if torch.cuda.is_available() else "cpu"
        self.model = Model.from_pretrained(model_name).to(self.device)
        self.sample_rate = 16000  # Standard sample rate for speaker embedding models
    
    async def extract_embeddings(
        self, 
        audio_path: str, 
        segments: List[Dict]
    ) -> List[Dict]:
        """Extract speaker embeddings for each audio segment."""
        try:
            # Load and preprocess audio
            waveform, sample_rate = torchaudio.load(audio_path)
            if sample_rate != self.sample_rate:
                resampler = torchaudio.transforms.Resample(
                    orig_freq=sample_rate, 
                    new_freq=self.sample_rate
                )
                waveform = resampler(waveform)
            
            embeddings = []
            for segment in segments:
                # Extract segment
                start_sample = int(segment['start'] * self.sample_rate)
                end_sample = int(segment['end'] * self.sample_rate)
                segment_waveform = waveform[:, start_sample:end_sample]
                
                # Extract embedding
                with torch.no_grad():
                    embedding = self.model(segment_waveform.to(self.device))
                
                embeddings.append({
                    'segment_id': segment['id'],
                    'start': segment['start'],
                    'end': segment['end'],
                    'embedding': embedding.cpu().numpy().tolist(),
                    'confidence': segment.get('confidence', 1.0)
                })
            
            return embeddings
            
        except Exception as e:
            logger.error(f"Error extracting speaker embeddings: {str(e)}")
            raise
    
    async def cluster_speakers(
        self, 
        embeddings: List[np.ndarray], 
        threshold: float = 0.7
    ) -> List[int]:
        """Cluster embeddings into speaker groups."""
        from sklearn.cluster import AgglomerativeClustering
        
        if not embeddings:
            return []
            
        # Convert to numpy array if needed
        if not isinstance(embeddings, np.ndarray):
            embeddings = np.array(embeddings)
        
        # Handle single segment case
        if len(embeddings) == 1:
            return [0]
            
        # Perform clustering
        clustering = AgglomerativeClustering(
            n_clusters=None,
            affinity='cosine',
            linkage='average',
            distance_threshold=1.0 - threshold  # Convert similarity to distance
        ).fit(embeddings)
        
        return clustering.labels_.tolist()
```

### 3. Speaker Matching Service

```python
class SpeakerMatchingService:
    def __init__(self, db: Session, threshold: float = 0.85):
        self.db = db
        self.threshold = threshold
    
    async def find_matching_speaker(
        self, 
        embedding: List[float],
        user_id: int,
        min_confidence: float = 0.7
    ) -> Optional[Dict]:
        """Find the best matching speaker for the given embedding."""
        # Convert to numpy array for calculations
        query_embedding = np.array(embedding, dtype=np.float32)
        
        # Get all speaker embeddings for this user
        speaker_embeddings = self.db.query(SpeakerEmbedding).join(Speaker).filter(
            Speaker.user_id == user_id,
            Speaker.is_active == True
        ).all()
        
        if not speaker_embeddings:
            return None
            
        # Calculate similarities
        similarities = []
        for spk_emb in speaker_embeddings:
            db_embedding = np.array(spk_emb.embedding, dtype=np.float32)
            similarity = self._cosine_similarity(query_embedding, db_embedding)
            similarities.append({
                'speaker_id': spk_emb.speaker_id,
                'similarity': similarity,
                'embedding_id': spk_emb.id
            })
        
        # Sort by similarity (highest first)
        similarities.sort(key=lambda x: x['similarity'], reverse=True)
        
        # Return best match if above threshold
        if similarities and similarities[0]['similarity'] >= self.threshold:
            return similarities[0]
            
        return None
    
    def _cosine_similarity(self, a: np.ndarray, b: np.ndarray) -> float:
        """Calculate cosine similarity between two vectors."""
        return np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b))
    
    async def create_new_speaker(
        self,
        name: str,
        user_id: int,
        embedding: List[float],
        metadata: Optional[Dict] = None
    ) -> Dict:
        """Create a new speaker with the given embedding."""
        try:
            async with self.db.begin():
                # Create speaker
                speaker = Speaker(
                    name=name,
                    user_id=user_id,
                    metadata=metadata or {}
                )
                self.db.add(speaker)
                await self.db.flush()
                
                # Save embedding
                speaker_embedding = SpeakerEmbedding(
                    speaker_id=speaker.id,
                    embedding=embedding,
                    confidence=1.0,
                    metadata={}
                )
                self.db.add(speaker_embedding)
                
                await self.db.commit()
                
                return {
                    'id': speaker.id,
                    'name': speaker.name,
                    'embedding_id': speaker_embedding.id
                }
                
        except Exception as e:
            await self.db.rollback()
            logger.error(f"Error creating new speaker: {str(e)}")
            raise
```

#### POST /api/speakers
```python
@router.post("/speakers", response_model=SpeakerResponse)
async def create_speaker(
    speaker: SpeakerCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create a new speaker profile."""
    service = SpeakerMatchingService(db)
    return await service.create_new_speaker(
        name=speaker.name,
        user_id=current_user.id,
        embedding=speaker.embedding,
        metadata=speaker.metadata
    )
```

#### GET /api/speakers/{speaker_id}/match
```python
@router.get("/speakers/{speaker_id}/match", response_model=List[MatchResult])
async def find_similar_speakers(
    speaker_id: int,
    threshold: float = Query(0.8, ge=0.5, le=1.0),
    limit: int = Query(10, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Find similar speakers based on voice characteristics."""
    # Get the target speaker's embeddings
    target_embeddings = await db.query(SpeakerEmbedding).filter(
        SpeakerEmbedding.speaker_id == speaker_id,
        Speaker.user_id == current_user.id
    ).all()
    
    if not target_embeddings:
        raise HTTPException(404, "Speaker not found")
    
    # Get all other speakers' embeddings
    other_embeddings = await db.query(SpeakerEmbedding).join(Speaker).filter(
        SpeakerEmbedding.speaker_id != speaker_id,
        Speaker.user_id == current_user.id,
        Speaker.is_active == True
    ).all()
    
    # Calculate similarities
    matches = []
    for other in other_embeddings:
        similarity = max(
            self._cosine_similarity(
                np.array(target.embedding, dtype=np.float32),
                np.array(other.embedding, dtype=np.float32)
            )
            for target in target_embeddings
        )
        
        if similarity >= threshold:
            matches.append({
                'speaker_id': other.speaker_id,
                'similarity': similarity,
                'embedding_id': other.id
            })
    
    # Sort and limit results
    matches.sort(key=lambda x: x['similarity'], reverse=True)
    return matches[:limit]
```

### 5. Integration with Transcription Pipeline

```python
class TranscriptionService:
    def __init__(self, db: Session):
        self.db = db
        self.whisperx = WhisperXService()
        self.speaker_embedding = SpeakerEmbeddingService()
        self.speaker_matching = SpeakerMatchingService(db)
    
    async def process_audio(self, file_path: str, user_id: int):
        # 1. Transcribe with WhisperX
        transcription = await self.whisperx.transcribe(file_path)
        
        # 2. Extract speaker embeddings
        segments = transcription['segments']
        embeddings = await self.speaker_embedding.extract_embeddings(file_path, segments)
        
        # 3. Cluster segments by speaker
        embedding_vectors = [e['embedding'] for e in embeddings]
        speaker_labels = await self.speaker_embedding.cluster_speakers(embedding_vectors)
        
        # 4. Match or create speakers
        speaker_map = {}
        for i, (segment, label) in enumerate(zip(segments, speaker_labels)):
            if label not in speaker_map:
                # Try to find matching speaker
                match = await self.speaker_matching.find_matching_speaker(
                    embedding=embedding_vectors[i],
                    user_id=user_id
                )
                
                if match:
                    speaker_map[label] = match['speaker_id']
                else:
                    # Create new speaker
                    speaker = await self.speaker_matching.create_new_speaker(
                        name=f"Speaker {label + 1}",
                        user_id=user_id,
                        embedding=embedding_vectors[i]
                    )
                    speaker_map[label] = speaker['id']
            
            # Add speaker ID to segment
            segment['speaker_id'] = speaker_map[label]
        
        return {
            'segments': segments,
            'speakers': [
                {'id': v, 'name': f"Speaker {k+1}"} 
                for k, v in speaker_map.items()
            ]
        }
```

### Phase 1: Core Infrastructure (1 week)
1. Set up pgvector extension in PostgreSQL
2. Create database schema for speakers and embeddings
3. Implement embedding extraction service
4. Add basic speaker management API

### Phase 2: Speaker Matching (1 week)
1. Implement similarity search
2. Add clustering for new speakers
3. Create matching service
4. Add API endpoints for speaker operations

### Phase 3: Integration (3 days)
1. Update transcription pipeline
2. Add speaker identification to processing
3. Implement batch processing for existing content

### Phase 4: UI & Testing (4 days)
1. Create speaker management interface
2. Add speaker assignment controls
3. Implement testing framework
4. Performance optimization

## Performance Considerations

1. **Vector Search**
   - Use IVFFlat index for approximate nearest neighbor search
   - Consider HNSW for higher accuracy with larger datasets
   - Implement caching for frequent queries

2. **Batch Processing**
   - Process multiple segments in parallel
   - Use bulk inserts for database operations
   - Implement background tasks for non-critical operations

3. **Resource Management**
   - Monitor GPU memory usage
   - Implement rate limiting
   - Add circuit breakers for external services

## Security Considerations

1. **Data Privacy**
   - Encrypt embeddings at rest
   - Implement access controls
   - Add audit logging

2. **API Security**
   - Rate limiting
   - Input validation
   - Authentication/authorization

## Monitoring and Metrics

1. **Performance Metrics**
   - Embedding extraction time
   - Search latency
   - Memory usage

2. **Quality Metrics**
   - Speaker clustering accuracy
   - False positive/negative rates
   - User correction rates

## Future Enhancements

1. **Active Learning**
   - Collect user corrections to improve models
   - Implement feedback loops

2. **Advanced Features**
   - Speaker diarization with overlap detection
   - Multi-lingual support
   - Emotion/sentiment analysis

3. **Integration**
   - Export/import speaker profiles
   - API for external systems
   - Webhook notifications
